Mathlib/Lean/Meta/RefinedDiscrTree/Pi.lean

```diff
@@ -1,115 +1,269 @@
-  if args.size ≤ 6 then
-    etaExpand args type lambdas 6 fun args lambdas =>
-      distributeLambdas lambdas type args[4]! args[5]!
+    k args lambdas (by omega)
+termination_by arity - args.size
+
+@[inline] private def app6? (fName : Name) (e : Expr) : Option (Expr × Expr × Expr) :=
+  if e.isAppOfArity fName 6 then
+    some (e.appFn!.appFn!.appArg!, e.appFn!.appArg!, e.appArg!)
   else
-  /- use that `(f + g) a = f a + g a` -/
-  let mut type := type
-  let mut inst := inst
-  let mut lhs := args[4]!
-  let mut rhs := args[5]!
-  for arg in args[6:] do
-    let mkApp3 (.const i _) _ f inst' := inst | return (type, lhs, rhs, lambdas)
-    unless i == instPi do return (type, lhs, rhs, lambdas)
-    type := .app f arg
-    inst := inst'
-    lhs := .app lhs arg
-    rhs := .app rhs arg
-  distributeLambdas lambdas type lhs rhs
-where
-  /-- use that `(fun x => f x + g x) = f + g` -/
-  distributeLambdas (lambdas : List FVarId) (type lhs rhs : Expr) :
-    MetaM (Expr × Expr × Expr × List FVarId) := match lambdas with
-    | fvarId :: lambdas => do
-      let decl ← fvarId.getDecl
-      let type := .forallE decl.userName decl.type (type.abstract #[.fvar fvarId]) decl.binderInfo
-      let lhs  := .lam     decl.userName decl.type (lhs.abstract  #[.fvar fvarId]) decl.binderInfo
-      let rhs  := .lam     decl.userName decl.type (rhs.abstract  #[.fvar fvarId]) decl.binderInfo
-      distributeLambdas lambdas type lhs rhs
-    | [] => return (type, lhs, rhs, [])
-
-/-- Normalize an application if the head is  `+`, `*`, `-` or `/`.
-Optionally return the `(type, lhs, rhs, lambdas)`. -/
-@[inline] def reduceHBinOp (n : Name) (args : Array Expr) (lambdas : List FVarId) :
-    MetaM (Option (Expr × Expr × Expr × List FVarId)) :=
-  match n with
-  | ``HAdd.hAdd => reduceHBinOpAux args lambdas ``instHAdd ``Pi.instAdd
-  | ``HMul.hMul => reduceHBinOpAux args lambdas ``instHMul ``Pi.instMul
-  | ``HSub.hSub => reduceHBinOpAux args lambdas ``instHSub ``Pi.instSub
-  | ``HDiv.hDiv => reduceHBinOpAux args lambdas ``instHDiv ``Pi.instDiv
-  | _ => return none
-
-/-- Normalize an application of a unary operator like `Inv.inv`, using:
-- `f⁻¹ a = (f a)⁻¹` to decrease the arity to 3
-- `(fun x => (f a)⁻¹) = f⁻¹` to get rid of any lambdas in front -/
-def reduceUnOpAux (args : Array Expr) (lambdas : List FVarId) (instPi : Name) :
-    OptionT MetaM (Expr × Expr × List FVarId) := do
-  guard (args.size ≥ 3)
-  let mut type := args[0]!
-  let mut inst := args[1]!
-  let mut arg := args[2]!
-  if args.size == 3 then
-    distributeLambdas lambdas type arg
+    none
+
+private def lambdaBody? (n : Nat) (e : Expr) : Option Expr :=
+match n, e with
+| 0, e => some e
+| n+1, .lam _ _ b _ => lambdaBody? n b
+| _, _ => none
+
+/-- Normalize an application if the head is `⁻¹` or `-`. -/
+private def reduceUnOp (n : Name) (args : Array Expr) (lambdas : List FVarId) :
+    MetaM (Option (List (Array (Option Expr) × List FVarId))) := withDefault do
```

Comment:
`Option (List (Array (Option Expr) × List FVarId` appears 4 times in this file; can you perhaps replace it with a `structure` so that you can document what the pieces mean?